src/robot/running/arguments/typeconverters.py

```diff
@@ -389,61 +389,165 @@ class ListConverter(TypeConverter):
+                if converter:
+                    types = ', '.join([type_name(t) for t in self.nested_types])
+                    tmp_list[i] = converter.convert(f"Tuple[{types}]", elem,
+                                                    explicit_type)
+            converted_tuple = tuple(tmp_list)
+        return converted_tuple
+
+    _convert = _non_string_convert
 
 @TypeConverter.register
 class DictionaryConverter(TypeConverter):
     type = dict
     abc = Mapping
+    abc = (Mapping, MutableMapping)
     type_name = 'dictionary'
-    aliases = ('dict', 'map')
+    aliases = ('dict', 'dictionary', 'map')
     value_types = (str, Mapping)
 
-    def _non_string_convert(self, value, explicit_type=True):
-        if issubclass(self.used_type, dict) and not isinstance(value, dict):
-            return dict(value)
-        return value
+    def __init__(self, type_, custom_converters=None, languages=None):
+        super().__init__(type_, languages=languages)
+        self.nested_types = getattr(type_, '__args__', ())
 
-    def _convert(self, value, explicit_type=True):
-        return self._literal_eval(value, dict)
+    def no_conversion_needed(self, value):
+        return self.used_type in self.abc and issubclass(type(value), self.abc)
+
+    def _non_string_convert(self, value, explicit_type=True):
+        if isinstance(value, str):
+            converted_dict = self._literal_eval(value, dict)
+        elif is_dict_like(value):
+            converted_dict = dict(value)
+        else:
+            raise ValueError
+        if self.nested_types and len(self.nested_types) == 2:
+            for key, elem in converted_dict.copy().items():
+                key_type = self.nested_types[0]
+                value_type = self.nested_types[1]
+                d_info = f"Dict[{type_name(key_type)}, {type_name(value_type)}]"
+                key_converter  = TypeConverter.converter_for(key_type,
+                                                             languages=self.languages)
+                elem_converter = TypeConverter.converter_for(value_type,
+                                                             languages=self.languages)
+                if key_converter:
+                    converted_key = key_converter.convert(f"key for {d_info}", key,
+                                                          explicit_type)
+                else:
+                    converted_key  = key
+                if elem_converter:
+                    converted_elem = elem_converter.convert(d_info, elem, explicit_type)
+                else:
+                    converted_elem = elem
+                replace_elem = type(converted_elem) != type(elem) 
+                if converted_key != key:
+                    converted_dict.pop(key)
+                    replace_elem = True
+                if replace_elem:
+                    converted_dict[converted_key] = converted_elem
```

Comment:
Modifying existing dictionary looks pretty complicated. Why not start with an empty dictionary that is populated based on (possibly) converted keys and values? Other converters would probably benefit from that approach as well.

Reply:
agreed